src/window/event_time_clock.rs

```diff
@@ -0,0 +1,328 @@
+use std::task::Poll;
+
+use chrono::{DateTime, Utc};
+use pyo3::{exceptions::PyValueError, pyclass, pymethods, Py, PyAny, PyResult, Python};
+
+use crate::{
+    pyo3_extensions::{TdPyAny, TdPyCallable},
+    recovery::StateBytes,
+    window::ClockConfig,
+};
+
+use super::{
+    system_clock::SystemClock,
+    testing_clock::{PyTestingClock, TestingClock, TestingClockConfig},
+    Builder, Clock, ClockBuilder,
+};
+
+/// Use datetimes from events as clock.
+///
+/// If the dataflow has no more input, all windows are closed.
+///
+/// The watermark is the system time since the last element
+/// plus the value of `late` plus the delay of the latest received element.
+/// It is updated every time an event with a newer datetime is processed.
+///
+/// Args:
+///
+///   dt_getter: Python function to get a datetime from an event.
+///
+///   wait_for_system_duration: How much (system) time to wait before considering an event late.
+///
+///   system_clock: Optional configuration to use a PyTestingClock as the internal system_clock,
+///                 only used for testing purposes right now.
+///
+/// Returns:
+///   Config object. Pass this as the `clock_config` parameter to your
+///   windowing operator.
+#[pyclass(module="bytewax.window", extends=ClockConfig)]
+#[pyo3(text_signature = "(dt_getter, wait_for_system_duration, system_clock)")]
```

Comment:
I think `wait_for_system_duration` is explicit but perhaps leaking implementation. What about something like `lateness_buffer` or `late_after_duration` or something?

Reply:
Yes, it's "internal implementation" but it's also the correct interpretation of this duration, as there's sort of multiple types of time going on.